docs test in docker (#360)
microfleet/core: packages/core/docs/recipes/testing.md

# Execute tests using Docker
Usually, any project you develop requires Unit and Integrational tests.
This recipe intended for Newbies, so if you are one of the Guru's that have their boilerplate, you can skip this document.

Not every test you develop could be run on your local PC because they require additional Software such as Database Brokers or other services.

> These recipes do not attempt to become the Full Solution and try to show some simple appliances.

Reviewer: These recipe is not intended to provide the full solution, just highlights the basic usecase

WARNING: The DEBUG variable is not set. Defaulting to a blank string.
yarn run v1.19.1
$ mocha

  server
    ✓ should be able to start (426ms)
    ✓ should say hello world (160ms)

  2 passing (592ms)

Done in 1.10s.
```

HOORAY! Your tests run in Docker.

## 4. Update your package.json
Now you can save the previous command in your `package.json`.
```json
{
  "scripts": {
    "test:docker": "docker-compose -f test/docker-compose.yml run --rm tester yarn test"
  },
}
```
After this, you will be able to run your tests using the `yarn test:docker` command.

# `mdep` As a Swedish Knife
`Microfleet` framework provides a special tool for instrumenting your development process. Please read [this page](https://www.npmjs.com/package/@makeomatic/deploy) to find advanced information and CLI commands.

## 1. Install `mdep` binary
First of all, you need to install the `@makeomatic/deploy` package, which provides `mdep` CLI command.

```console
u:~$ npm install --save @makeomatic/deploy
u:~$ npm install --save-dev cross-env
u:~$ npm install --save-dev nyc
u:~$ npm install --save-dev @babel/core @babel/register babel-plugin-istanbul babel-plugin-transform-strict-mode
```

> The package adds `husky` and `semantic-release` packages into your `package.json` and adds Git Commit message validation hook.

Reviewer: ```suggestion
The package adds `husky` and `semantic-release` packages into your `package.json` and adds Git commit message validation hook.
```

Now you can check that installation is successful by running:
```console
u:~$ npx mdep --help
mdep <command>

Commands:
  mdep docker          manages docker lifecycle
  mdep test <command>  performs tests in docker

Options:
  --node, -n                  node version to use when building
                                                            [default: "12.13.0"]
  --env, -E                   node environment to build for
                                                         [default: "production"]
  --project, -p               project name where this is used
                                                [default: "microfleet-demo-app"]
  --docker_build_args, --dba  docker build args
  --repository, --repo        docker repository to use   [default: "makeomatic"]
  --version, -v               version of the project to build [default: "1.0.0"]
  --pkg                       package json path

>   [default: "/home/pajgo/Desktop/makeomatic/microfleet/create-microfleet-app/pac

Reviewer: depersonalize the path please (just in terms of consistency: we've decided to hide such artifacts few pull requests ago)

                                                                     kage.json"]
  --help                      Show help                                [boolean]

```

## 2. Init Test environment

> After `mdep` tool installed, you can initialize your test environment with `babel`, `mocha` and `docker-compose`:

Reviewer: ```suggestion
After the `mdep` tool is installed, you can initialize your test environment with `babel`, `mocha` and `docker-compose`:
```

```console
u:~$ npx mdep test init
```
Now if you check your source tree you will see that generic test environment created:

```console
u:~$ git status
...
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	.babelrc                  - Babel configuration
	.nycrc                    - Istanbul Nyc code coverage configuration
	.releaserc.json           - Semantic Release configuration
	test/docker-compose.yml   - docker-compose configuration
	test/mocha.opts           - Mocha configuration
```

> The `docker-compose.yml` file already has one service defined. This service will be used to run your tests using the `mdep test run` command. Now you can check that your tests can run in a Dockerized environment using Node.js image defined inside of your `test/docker-compose.yml`. You can find all available `Tester` images on [this page](https://hub.docker.com/r/makeomatic/node/tags).

Reviewer: ```suggestion
The `docker-compose.yml` file already has one service defined. This service is going to be used to run your tests using the `mdep test run` command. Now you can check that your tests can run in a Dockerized environment using Node.js image defined inside of your `test/docker-compose.yml`. You can find all available `Tester` images on [this page](https://hub.docker.com/r/makeomatic/node/tags).
```

Reviewer: but I'm not sure
=)
